pycozmo/window.py

```diff
@@ -5,159 +5,167 @@
 """
 
 import math
-from typing import Optional, Any
+from typing import Optional, List, Tuple, Any
+
+from . import exception
 
 
 __all__ = [
     "BaseWindow",
     "ReceiveWindow",
-    "SendWindowSlot",
     "SendWindow",
 ]
 
 
 class BaseWindow(object):
+    """ Base communication window class. """
 
-    def __init__(self, seq_bits: int, size: Optional[int] = None) -> None:
+    def __init__(self, seq_bits: int, size: Optional[int] = None, max_seq: Optional[int] = None) -> None:
+        """ Crate a window by specifying either sequence number bits or size of the window. """
         if size is None:
-            self.size = int(math.pow(2, seq_bits - 1))
-        elif size < 0 or size > int(math.pow(2, seq_bits - 1)):
+            if seq_bits < 1:
+                raise ValueError("Invalid sequence number bits.")
+            size = int(math.pow(2, seq_bits - 1))
+        elif size <= 0 or size > int(math.pow(2, seq_bits - 1)):
             raise ValueError("Invalid window size.")
-        else:
-            self.size = size
+        # Size of the window.
+        self.size = size
+        # Next expected sequence number (0, max_seq-1).
         self.expected_seq = 1
-        self.last_seq = 0
-        self.max_seq = int(math.pow(2, seq_bits)) - 1
+        # Maximum sequence number (first invalid).
+        self.max_seq = max_seq or int(math.pow(2, seq_bits))
+        if self.max_seq % self.size != 0:
+            raise ValueError("The maximum sequence number must be a multiple of the window size must.")
+        # Window data
+        self.window = [None for _ in range(self.size)]
 
     def is_valid_seq(self, seq: int) -> bool:
+        """ Check whether a sequence number is valid for the window. """
         res = 0 <= seq < self.max_seq
         return res
 
     def reset(self) -> None:
+        """ Reset the window. """
         self.expected_seq = 1
-        self.last_seq = 0
+        self.window = [None for _ in range(self.size)]
 
 
 class ReceiveWindow(BaseWindow):
+    """
+    Receive communication window class.
 
-    def __init__(self, seq_bits: int, size: Optional[int] = None) -> None:
-        super().__init__(seq_bits, size)
-        self.window = [None for _ in range(self.size)]
+    When packets are received (in whatever order), they are put in the window using the put() method.
+
+    Packets are extracted from the window in the expected order using the get() method.
+    """
+
+    def __init__(self, seq_bits: int, size: Optional[int] = None, max_seq: Optional[int] = None) -> None:
+        """ Crate a window by specifying either sequence number bits or size of the window. """
+        super().__init__(seq_bits, size, max_seq)
+        # Last used sequence number (0, max_seq-1).
+        self.last_seq = (self.expected_seq + self.size - 1) % self.max_seq
```

Comment:
Maybe we are missing an `expected_seq` initialization parameter in this init? Otherwise, this operation will always result in the same value: `self.size`, so we could skip the calculation.

I also don't know if this operation is correct. `self.size` corresponds to the window size, while the sequences work in a different space: the window rolls over the sequence values.
Then, if I understand correctly, the initialization should be something like this, maybe?
```
self.expected_seq = expected_seq if expected_seq is not None else 1
self.last_seq = (self.expected_seq - 1) % self.max_seq
```

Reply:
No need to pass `expected_seq` as an argument, as the protocol always bootstraps communication, starting with 1.

No need to initialize `self.expected_seq` as the base class initializes it to 1.

The base class, also initializes `self.size`.

I introduced a new requirement that the window `size` is a factor of `max_seq`, so the calculation seems correct. Can you come up with a test case that makes the class behave incorrectly?

Reply:
Sorry, you are right! The name confused me, I thought last_seq was the last received sequence, and not the last one that would fit after the expected one, the logic is correct.

In any case, we can simplify this to be:
`self.last_seq = self.size`.

If I understand correctly, on initialization we will always have:
```
self.expected_seq = 1
self.size = (whatever size)
```
Then, `self.last_seq` = `( 1 + size - 1) % self.max_seq` = `size` if `size < max_seq`. 
I believe that for the protocol to work the size of the window should be always equal or less than half of the index range (`max_seq`), therefore `size` will always be less than `max_seq`.

Maybe we could include a check to ensure (and remember) that the window size should never have less than half the size of the sequence range also?

Reply:
There is already a check actually - https://github.com/zayfod/pycozmo/blob/harden-window/pycozmo/window.py#L29

Reply:
All good then :) Sorry about the confusion!